File: PracticasProgramacion/Practica_18/python/ordenador_merge_sort.py
"""
OrdenadorMergeSort: divide y conquista. Merge de mitades ordenadas.
O(n log n). Soporta char, int, AutoPOO, AutoPE, PersonaPOO, PersonaPE.
"""

from dataclasses import dataclass
from typing import List


@dataclass
class AutoPE:
    precio: float
    anio: int


@dataclass
class PersonaPE:
    nombre: str
    edad: int


class AutoPOO:
    def __init__(self, precio: float = 0, anio: int = 0) -> None:
        self._precio = precio
        self._anio = anio

    def get_precio(self) -> float:
        return self._precio


class PersonaPOO:
    def __init__(self, nombre: str = "", edad: int = 0) -> None:
        self._nombre = nombre
        self._edad = edad

    def get_nombre(self) -> str:
        return self._nombre


class OrdenadorMergeSort:
    def ordenar_caracteres(self, arr: List[str]) -> None:
        if len(arr) <= 1:
            return
        mid = len(arr) // 2
        L, R = arr[:mid], arr[mid:]
        self.ordenar_caracteres(L)
        self.ordenar_caracteres(R)
        i = j = k = 0
        while i < len(L) and j < len(R):
            if L[i] <= R[j]:
                arr[k] = L[i]
                i += 1
            else:
                arr[k] = R[j]
                j += 1
            k += 1
        arr[k:] = L[i:] or R[j:]

    def ordenar_enteros(self, arr: List[int]) -> None:
        if len(arr) <= 1:
            return
        mid = len(arr) // 2
        L, R = arr[:mid], arr[mid:]
        self.ordenar_enteros(L)
        self.ordenar_enteros(R)
        i = j = k = 0
        while i < len(L) and j < len(R):
            if L[i] <= R[j]:
                arr[k] = L[i]
                i += 1
            else:
                arr[k] = R[j]
                j += 1
            k += 1
        arr[k:] = L[i:] or R[j:]

    def ordenar_autos(self, arr: List[AutoPOO]) -> None:
        if len(arr) <= 1:
            return
        mid = len(arr) // 2
        L, R = arr[:mid], arr[mid:]
        self.ordenar_autos(L)
        self.ordenar_autos(R)
        i = j = k = 0
        while i < len(L) and j < len(R):
            if L[i].get_precio() <= R[j].get_precio():
                arr[k] = L[i]
                i += 1
            else:
                arr[k] = R[j]
                j += 1
            k += 1
        arr[k:] = L[i:] or R[j:]

    def ordenar_autos_pe(self, arr: List[AutoPE]) -> None:
        arr.sort(key=lambda x: x.precio)

    def ordenar_personas(self, arr: List[PersonaPOO]) -> None:
        if len(arr) <= 1:
            return
        mid = len(arr) // 2
        L, R = arr[:mid], arr[mid:]
        self.ordenar_personas(L)
        self.ordenar_personas(R)
        i = j = k = 0
        while i < len(L) and j < len(R):
            if L[i].get_nombre() <= R[j].get_nombre():
                arr[k] = L[i]
                i += 1
            else:
                arr[k] = R[j]
                j += 1
            k += 1
        arr[k:] = L[i:] or R[j:]

    def ordenar_personas_pe(self, arr: List[PersonaPE]) -> None:
        arr.sort(key=lambda x: x.nombre)

File: PracticasProgramacion/Practica_18/cpp/OrdenadorMergeSort.cpp
#include "OrdenadorMergeSort.h"
#include <algorithm>
#include <cstring>

void OrdenadorMergeSort::merge(char arr[], int l, int m, int r) {
    int n1 = m - l + 1, n2 = r - m;
    char* L = new char[n1], * R = new char[n2];
    for (int i = 0; i < n1; ++i) L[i] = arr[l + i];
    for (int j = 0; j < n2; ++j) R[j] = arr[m + 1 + j];
    int i = 0, j = 0, k = l;
    while (i < n1 && j < n2)
        arr[k++] = (L[i] <= R[j]) ? L[i++] : R[j++];
    while (i < n1) arr[k++] = L[i++];
    while (j < n2) arr[k++] = R[j++];
    delete[] L; delete[] R;
}
void OrdenadorMergeSort::mergeSort(char arr[], int l, int r) {
    if (l >= r) return;
    int m = l + (r - l) / 2;
    mergeSort(arr, l, m);
    mergeSort(arr, m + 1, r);
    merge(arr, l, m, r);
}
void OrdenadorMergeSort::ordenarCaracteres(char arr[], int n) { mergeSort(arr, 0, n - 1); }

void OrdenadorMergeSort::merge(int arr[], int l, int m, int r) {
    int n1 = m - l + 1, n2 = r - m;
    int* L = new int[n1], * R = new int[n2];
    for (int i = 0; i < n1; ++i) L[i] = arr[l + i];
    for (int j = 0; j < n2; ++j) R[j] = arr[m + 1 + j];
    int i = 0, j = 0, k = l;
    while (i < n1 && j < n2)
        arr[k++] = (L[i] <= R[j]) ? L[i++] : R[j++];
    while (i < n1) arr[k++] = L[i++];
    while (j < n2) arr[k++] = R[j++];
    delete[] L; delete[] R;
}
void OrdenadorMergeSort::mergeSort(int arr[], int l, int r) {
    if (l >= r) return;
    int m = l + (r - l) / 2;
    mergeSort(arr, l, m);
    mergeSort(arr, m + 1, r);
    merge(arr, l, m, r);
}
void OrdenadorMergeSort::ordenarEnteros(int arr[], int n) { mergeSort(arr, 0, n - 1); }

void OrdenadorMergeSort::merge(AutoPOO arr[], int l, int m, int r) {
    int n1 = m - l + 1, n2 = r - m;
    AutoPOO* L = new AutoPOO[n1], * R = new AutoPOO[n2];
    for (int i = 0; i < n1; ++i) L[i] = arr[l + i];
    for (int j = 0; j < n2; ++j) R[j] = arr[m + 1 + j];
    int i = 0, j = 0, k = l;
    while (i < n1 && j < n2)
        arr[k++] = (L[i].getPrecio() <= R[j].getPrecio()) ? L[i++] : R[j++];
    while (i < n1) arr[k++] = L[i++];
    while (j < n2) arr[k++] = R[j++];
    delete[] L; delete[] R;
}
void OrdenadorMergeSort::mergeSort(AutoPOO arr[], int l, int r) {
    if (l >= r) return;
    int m = l + (r - l) / 2;
    mergeSort(arr, l, m);
    mergeSort(arr, m + 1, r);
    merge(arr, l, m, r);
}
void OrdenadorMergeSort::ordenarAutos(AutoPOO arr[], int n) { mergeSort(arr, 0, n - 1); }

void OrdenadorMergeSort::merge(AutoPE arr[], int l, int m, int r) {
    int n1 = m - l + 1, n2 = r - m;
    AutoPE* L = new AutoPE[n1], * R = new AutoPE[n2];
    for (int i = 0; i < n1; ++i) L[i] = arr[l + i];
    for (int j = 0; j < n2; ++j) R[j] = arr[m + 1 + j];
    int i = 0, j = 0, k = l;
    while (i < n1 && j < n2)
        arr[k++] = (L[i].precio <= R[j].precio) ? L[i++] : R[j++];
    while (i < n1) arr[k++] = L[i++];
    while (j < n2) arr[k++] = R[j++];
    delete[] L; delete[] R;
}
void OrdenadorMergeSort::mergeSort(AutoPE arr[], int l, int r) {
    if (l >= r) return;
    int m = l + (r - l) / 2;
    mergeSort(arr, l, m);
    mergeSort(arr, m + 1, r);
    merge(arr, l, m, r);
}
void OrdenadorMergeSort::ordenarAutosPE(AutoPE arr[], int n) { mergeSort(arr, 0, n - 1); }

void OrdenadorMergeSort::merge(PersonaPOO arr[], int l, int m, int r) {
    int n1 = m - l + 1, n2 = r - m;
    PersonaPOO* L = new PersonaPOO[n1], * R = new PersonaPOO[n2];
    for (int i = 0; i < n1; ++i) L[i] = arr[l + i];
    for (int j = 0; j < n2; ++j) R[j] = arr[m + 1 + j];
    int i = 0, j = 0, k = l;
    while (i < n1 && j < n2)
        arr[k++] = (L[i].getNombre() <= R[j].getNombre()) ? L[i++] : R[j++];
    while (i < n1) arr[k++] = L[i++];
    while (j < n2) arr[k++] = R[j++];
    delete[] L; delete[] R;
}
void OrdenadorMergeSort::mergeSort(PersonaPOO arr[], int l, int r) {
    if (l >= r) return;
    int m = l + (r - l) / 2;
    mergeSort(arr, l, m);
    mergeSort(arr, m + 1, r);
    merge(arr, l, m, r);
}
void OrdenadorMergeSort::ordenarPersonas(PersonaPOO arr[], int n) { mergeSort(arr, 0, n - 1); }

void OrdenadorMergeSort::merge(PersonaPE arr[], int l, int m, int r) {
    int n1 = m - l + 1, n2 = r - m;
    PersonaPE* L = new PersonaPE[n1], * R = new PersonaPE[n2];
    for (int i = 0; i < n1; ++i) L[i] = arr[l + i];
    for (int j = 0; j < n2; ++j) R[j] = arr[m + 1 + j];
    int i = 0, j = 0, k = l;
    while (i < n1 && j < n2)
        arr[k++] = (L[i].nombre <= R[j].nombre) ? L[i++] : R[j++];
    while (i < n1) arr[k++] = L[i++];
    while (j < n2) arr[k++] = R[j++];
    delete[] L; delete[] R;
}
void OrdenadorMergeSort::mergeSort(PersonaPE arr[], int l, int r) {
    if (l >= r) return;
    int m = l + (r - l) / 2;
    mergeSort(arr, l, m);
    mergeSort(arr, m + 1, r);
    merge(arr, l, m, r);
}
void OrdenadorMergeSort::ordenarPersonasPE(PersonaPE arr[], int n) { mergeSort(arr, 0, n - 1); }
